libcudacxx/include/cuda/std/__numeric/saturation_arithmetic.h

```diff
@@ -0,0 +1,850 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCUDACXX___NUMERIC_SATURATION_ARITHMETIC_H
+#define _LIBCUDACXX___NUMERIC_SATURATION_ARITHMETIC_H
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#if _CCCL_STD_VER >= 2017
+
+#  include <cuda/std/__algorithm/clamp.h>
+#  include <cuda/std/__algorithm/max.h>
+#  include <cuda/std/__algorithm/min.h>
+#  include <cuda/std/__concepts/concept_macros.h>
+#  include <cuda/std/__type_traits/is_arithmetic_integral.h>
+#  include <cuda/std/__type_traits/is_constant_evaluated.h>
+#  include <cuda/std/__type_traits/is_signed.h>
+#  include <cuda/std/__type_traits/is_unsigned.h>
+#  include <cuda/std/__type_traits/make_unsigned.h>
+#  include <cuda/std/__type_traits/remove_cv.h>
+#  include <cuda/std/__utility/cmp.h>
+#  include <cuda/std/climits>
+#  include <cuda/std/cstdint>
+#  include <cuda/std/limits>
+
+#  include <nv/target>
+
+#  if _CCCL_COMPILER(MSVC)
+#    include <intrin.h>
+#  endif // _CCCL_COMPILER(MSVC)
+
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+class __add_sat
+{
+  template <class _Tp>
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr _Tp
+  __clamp_overflow(_Tp, _Tp __y, _Tp __result, bool __overflow) noexcept
+  {
+    if (__overflow)
+    {
+      if constexpr (_CCCL_TRAIT(is_unsigned, _Tp))
+      {
+        __result = _CUDA_VSTD::numeric_limits<_Tp>::max();
+      }
+      else
+      {
+        __result = (__y < 0) ? _CUDA_VSTD::numeric_limits<_Tp>::min() : _CUDA_VSTD::numeric_limits<_Tp>::max();
+      }
+    }
+    return __result;
+  }
+
+  template <class _Tp>
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static _Tp __impl_generic(_Tp __x, _Tp __y) noexcept
```

Comment:
Is there a reason this is not `constexpr`?
```suggestion
  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr _Tp __impl_generic(_Tp __x, _Tp __y) noexcept
```

Reply:
There was no reason to make it `constexpr`, we have a separate `constexpr` implementation

Reply:
I've just realized we can just get rid of the constexpr implementation completely